### PR TITLE
Backup and restore stub

### PR DESCRIPTION
**Pull request by gsanchietti**
## Assumptions
- this is implementation is for disaster-recovery scenario: restore a module to a different node/cluster
- `restic` is the only engine
- all backups must be done using the network (no local storage)
- a target storage can contain multiple backup instances, each module instance has its own subdirectory to avoid conflicts
- every module instance should handle its own backup and restore and decide what to save (redis keys, podman volumes). The `state/environment` file is always included in the backup, as it is required by the `cluster/restore-module` action
- `rclone` can be used to list what backups are stored inside a remote backup storage

## Implementation details
- the `agent` Python package implements the common logic for backup and restore in the `agent.run_restic()` function
- `module-backup` and `module-restore` agent commands provide the basic implementations
- each module can implement `module-dump-state` and `module-cleanup-state` to prepare/cleanup the data that has to be included in the backup
- the restore is implemented using a `restore-module` action inside the module agent, each module can extend it to implement specific restore steps. The basic `10restore_module` actually invokes the `module-restore` command.
- backups are scheduled by Systemd timers

New module agent APIs:
- `configure-backup` manages Systemd timer and service units that actually start the module backup procedure
- `restore-module` implements restore (see details)
- `run-backup` starts backup procedure immediately

New cluster agent APIs:
- `list-backup-repositories`, `add-backup-repository`, `alter-backup-repository`, `remove-backup-repository`
- `read-backup-repository`
- `list-backups`,`add-backup`, `alter-backup`, `remove-backup`
- `run-backup`
- `restore-module`: executes `add-module` as subtask, then runs `restore-module` on the new instance

Core changes:
- added `MODULE_UUID` to module environment inside the `add-module`: this field can be used to identify an instance regardless of its `module_id` which can change during restore
- Defined PYTHONPATH as "${AGENT_INSTALL_DIR}/pypkg" (useful to ship additional Python packages)

## Usage examples

Create a Backblaze B2 repository (empty password implies the generation of a new random token used by restic to encrypt the backup data)

    api-cli run add-backup-repository --data '{"name":"BackBlaze repo1","url":"b2:backupns8","parameters":{"b2_account_id":"xxx","b2_account_key":"xxx"},"provider":"backblaze","password":""}'

Configure a backup of `dokuwiki1` and `samba1` instances using the same repository, with default `name` assignment:

    api-cli run add-backup --data '{"repository":"48ce000a-79b7-5fe6-8558-177fd70c27b4","schedule":"daily","retention":7,"instances":["dokuwiki1","samba1"],"enabled":true, "name":""}'
    1

Run backup 1 now:

    api-cli run run-backup --data '{"id":1}'

Execute a restore inside a new module instance:

    api-cli run cluster/restore-module --data '{"node":1, "repository":"48ce000a-79b7-5fe6-8558-177fd70c27b4", "path":"dokuwiki/dokuwiki1@3792c7db-9450-4bd3-84a3-034cd0087839","snapshot":""}'

List contents of a backup repository:

    api-cli run read-backup-repository --data '{"repository":"48ce000a-79b7-5fe6-8558-177fd70c27b4"}' | jq

## TODOs

- [x] validation for all APIs
- [x] rootfull and rootless restore
- [x] list of target storage to know which modules have a backup
- [x] move the backup password outside the module scope: the password should be available at cluster level because it's necessary to restore a module

## Future work

Backup/Restore of core data has not been included in this PR. 

Core backup could be restored from UI during the first configuration wizard.
It could include (not implemented):

- UI favorites
- User  ACLs
- Backup list
- List of installed applications
- External account provider configurations
- Repository list
- Other files with settings (eg. wireguard keys)

### CODE REVIEW

File: core/imageroot/var/lib/nethserver/cluster/actions/restore-module/50restore [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python3
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/restore-module/50restore:1]: Move to node_agent

File: core/imageroot/usr/local/agent/pypkg/agent/backup/__init__.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            cmd = ["restic", "--cache-dir", cache_dir, "-r", f"{self.config['url']}:{self.config['module']}"]
+            p_dump = None
+            env = dict()
+            # First try to access rootless env, if it fails, switch to rootfull one
+            try:
+                p_dump = subprocess.run(cmd + ["dump", "latest", rootless_env], env=restic_env, check=True, capture_output=True)
+            except:
+                p_dump = subprocess.run(cmd + ["dump", "latest", rootfool_env], env=restic_env, check=True, capture_output=True)
+            for line in p_dump.stdout.splitlines():
+                var, sep, val = line.decode().partition("=")
```
[the remark this answers is not in the document]
Comment (gsanchietti) [reply core/imageroot/usr/local/agent/pypkg/agent/backup/__init__.py:98]: decode seems mandatatory

File: core/imageroot/var/lib/nethserver/cluster/actions/add-backup-repository/50update [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# Setup extra parameters for repository access
+for k in request['parameters']:
+    data[k] = request['parameters'][k]
+
+# Add the repository key
+if not rdb.hset(f'cluster/backup_repository/{name}', mapping=data):
+    sys.exit(1)
+
+# Create repository path to store rclone configuration
+repo_path = f'/var/lib/nethserver/backup_repository/{name}'
```
[the remark this answers is not in the document]
Comment (gsanchietti) [reply core/imageroot/var/lib/nethserver/cluster/actions/add-backup-repository/50update:54]: Ok. What about restic cache dir? Currently is saved under `/var/lib/nethserver/cache/restic`.
Note that this directory can grow to hundreds of megabytes.

File: core/imageroot/usr/local/agent/pypkg/agent/backup/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavidePrincipi) [reply core/imageroot/usr/local/agent/pypkg/agent/backup/__init__.py]: It is mandatory if the subprocess stream is considered binary data as in your case. Look at the differences, with and without `encoding="utf-8"`...

```
[root@fc1 ~]# python -c 'import subprocess ; print(subprocess.run(["echo", "à"], capture_output=True))'
CompletedProcess(args=['echo', 'à'], returncode=0, stdout=b'\xc3\xa0\n', stderr=b'')
[root@fc1 ~]# python -c 'import subprocess ; print(subprocess.run(["echo", "à"], capture_output=True, encoding="utf-8"))'
CompletedProcess(args=['echo', 'à'], returncode=0, stdout='à\n', stderr='')
```

**Pull request merged**